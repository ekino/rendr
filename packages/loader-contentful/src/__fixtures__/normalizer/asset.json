{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "8VoLF5Jla5JcMnhNV74Oj",
        "type": "Asset",
        "createdAt": "2019-08-09T21:30:18.823Z",
        "updatedAt": "2019-08-09T21:30:18.823Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "locale": "en-US"
    },
    "fields": {
        "title": "test",
        "description": "test",
        "file": {
            "url": "//images.ctfassets.net/f5bx295b472q/8VoLF5Jla5JcMnhNV74Oj/4f3afcf599ffbcc55528e2026e40cd0d/Screenshot_2019-07-18_at_16.39.23.png",
            "details": {
                "size": 10072,
                "image": {
                    "width": 210,
                    "height": 134
                }
            },
            "fileName": "Screenshot 2019-07-18 at 16.39.23.png",
            "contentType": "image/png"
        }
    }
}