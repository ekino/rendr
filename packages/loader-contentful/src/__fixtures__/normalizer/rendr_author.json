{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "7pOMMe1TEojuNC9gKdxR0k",
        "type": "Entry",
        "createdAt": "2019-08-20T18:39:11.532Z",
        "updatedAt": "2019-08-20T18:39:11.532Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_author"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "name": "Thomas Rabaix",
        "job_title": "CTO",
        "slug": "thomas-rabaix",
        "social_twitter": "th0masr",
        "biography": "CTO at Ekino. Sonata Project founder. Symfony Expert. Lately doing python, go, ansible and cloud platforms."
    }
}