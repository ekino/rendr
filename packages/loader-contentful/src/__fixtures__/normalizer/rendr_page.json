{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "4QrVSXjiXz6iSS9Vli7sDh",
        "type": "Entry",
        "createdAt": "2019-08-20T18:26:10.936Z",
        "updatedAt": "2019-08-20T18:28:43.549Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 3,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_page"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "title": "Ekino Homepage",
        "seo_description": "Homepage",
        "seo_keywords": "digital, awesome",
        "extends": "root",
        "path": "/",
        "blocks": [
            {
                "sys": {
                    "space": {
                        "sys": {
                            "type": "Link",
                            "linkType": "Space",
                            "id": "f5bx295b472q"
                        }
                    },
                    "id": "1ovjOotrujpd5BTTKIxq8C",
                    "type": "Entry",
                    "createdAt": "2019-08-20T18:25:58.971Z",
                    "updatedAt": "2019-08-20T18:25:58.971Z",
                    "environment": {
                        "sys": {
                            "id": "master",
                            "type": "Link",
                            "linkType": "Environment"
                        }
                    },
                    "revision": 1,
                    "contentType": {
                        "sys": {
                            "type": "Link",
                            "linkType": "ContentType",
                            "id": "rendr_block_text"
                        }
                    },
                    "locale": "en-US"
                },
                "fields": {
                    "internal_title": "About section",
                    "container": "header",
                    "order": 1,
                    "title": "About",
                    "contents": "We create digital services that transform businesses.\n\nUsing lean, Agile methodologies and rapid prototyping, it’s never been easier to create, test, and refine new services based on customer data and human behaviour.\n\nTogether, we’ll collaborate to accelerate your innovation and increase your differentiation."
                }
            }
        ],
        "layout": "default",
        "published_at": "2019-08-16T00:00",
        "settings": {
            "foo": "bar"
        }
    }
}