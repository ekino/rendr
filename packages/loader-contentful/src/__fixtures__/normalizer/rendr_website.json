{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "75Vqk3IpfxlzQfvhdSASZV",
        "type": "Entry",
        "createdAt": "2019-08-20T18:50:08.851Z",
        "updatedAt": "2019-08-20T18:50:08.851Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_website"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "name": "Ekino.com",
        "domains": [
            "www.ekino.com"
        ],
        "path": "/",
        "culture": "en_GB",
        "country_code": "FRA",
        "order": 1,
        "enabled": true
    }
}