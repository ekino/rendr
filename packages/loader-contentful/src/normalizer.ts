import { Entry } from "contentful";
import {
  Page,
  Map,
  BlockDefinition,
  Settings,
  NormalizationError,
  createPage,
  RendrCtx,
} from "@ekino/rendr-core";

import {
  Asset,
  Website,
  ContentfulWebsite,
  ContentfulPage,
  ContentfulAsset,
  EntryNormalizerList,
  EntryNormalizer,
  BaseContentfulBlock,
} from "./types";

const ASSET_REGEXP = /(images\.(ctfassets\.net|contentful\.com)([0-9a-zA-Z/\-_]*)\.([a-z]{1,4}))/g;
const ASSET_FORMATS: Map = {
  jpg: "jpg",
  jpeg: "jpg",
  png: "png",
};

export const emptyPicture = {
  enabled: false,
  id: "-",
  title: "No picture",
  url: "",
};

export const defaultNormalizers: EntryNormalizerList = {
  rendr_page: normalizePage,
  rendr_website: normalizeWebsite,
  asset: normalizePicture, // Asset is the default from contentul
};

export function createNormalizer(
  extraNormalizers: EntryNormalizerList = {}
): EntryNormalizer {
  const normalizers: EntryNormalizerList = {
    ...defaultNormalizers,
    ...extraNormalizers,
  };

  // check for recursive loop, and protect against
  // infinite loop.
  const stack: string[] = [];

  return async function normalizer(
    ctx: RendrCtx,
    entry: Entry<any> | ContentfulAsset,
    normalizerFn?: EntryNormalizer
  ): Promise<any> {
    if (!validEntry(entry)) {
      return;
    }

    if (stack.includes(entry.sys.id)) {
      // recursive call cannot normalize
      return false;
    }

    let key;
    if (entry.sys.type == "Asset") {
      key = "asset";
    } else if (
      entry.sys.type == "Entry" &&
      entry.sys.contentType.sys.id in normalizers
    ) {
      key = entry.sys.contentType.sys.id;
    } else {
      // entry.sys.type == 'Link' if an entity is not resolvable
      // ie: Contentful API returns a `notResolvable` error.
      return;
    }

    stack.push(entry.sys.id);

    try {
      return await normalizers[key](ctx, entry, normalizer);
    } catch (err) {
      // log error
      if (err instanceof NormalizationError) {
        throw err;
      }

      throw new NormalizationError(err);
    } finally {
      stack.pop();
    }
  };
}

export function fixImageUrl(content: string) {
  if (!content) {
    return "";
  }

  return content.replace(ASSET_REGEXP, (replacement, ...params: string[]) => {
    if (params[3] in ASSET_FORMATS) {
      return `images.ctfassets.net${params[2]}.${params[3]}?fm=${
        ASSET_FORMATS[params[3]]
      }`;
    }

    return replacement;
  });
}

export function fixHttps(url: string) {
  if (url.length < 2) {
    return url;
  }

  if (url.substr(0, 2) == "//") {
    return `https:${url}`;
  }

  return url;
}

export function validEntry(
  entity: Entry<any> | ContentfulAsset,
  contentType = ""
) {
  if (!entity) {
    return false;
  }

  if (!("sys" in entity) || !("id" in entity.sys)) {
    return false;
  }

  if (contentType.length === 0) {
    return true;
  }

  return entity.sys.contentType.sys.id === contentType;
}

export async function normalizePage(
  ctx: RendrCtx,
  entry: Entry<ContentfulPage>,
  normalizer: EntryNormalizer
): Promise<Page> {
  const ttl = ctx.settings.rendr_site
    ? (ctx.settings.rendr_site as Website).cache.ttl
    : 0;
  const sharedTtl = ctx.settings.rendr_site
    ? (ctx.settings.rendr_site as Website).cache.sharedTtl
    : 0;

  const data = {
    id: entry.sys.id,
    head: {
      title: entry.fields.title,
      meta: [] as Settings[],
    },
    path: entry.fields.path ?? "",
    template: entry.fields.layout ?? "default",
    settings: entry.fields.settings,
    blocks: [] as BlockDefinition[],
    cache: {
      ttl: entry.fields.ttl ?? ttl,
      sharedTtl: entry.fields.shared_ttl ?? sharedTtl,
    },
  };

  const keys = Object.keys(entry.fields);

  for (let k in keys) {
    const key = keys[k];

    // NAME_blocks convention
    if (key.substr(-7) !== "_blocks") {
      continue;
    }

    // @ts-ignore
    const blocks = entry.fields[key] as Entry<any>[];
    const container = ((word: string) => {
      const words = word.split("_blocks");
      words.pop();

      return words.join();
    })(key);

    if (!blocks) {
      continue;
    }

    for (let index in blocks) {
      const block = blocks[index];

      const def = await normalizer(ctx, block);

      if (!def) {
        continue;
      }

      def.id = block.sys.id;
      def.container = container;
      def.order = index;

      data.blocks.push(def);
    }
  }

  if (entry.fields.seo_keywords && entry.fields.seo_keywords.length > 0) {
    data.head.meta.push({
      name: "keywords",
      content: entry.fields.seo_keywords,
    });
  }

  if (entry.fields.seo_description && entry.fields.seo_description.length > 0) {
    data.head.meta.push({
      name: "description",
      content: entry.fields.seo_description,
    });
  }

  return createPage(data);
}

export function normalizePicture(ctx: RendrCtx, entry: ContentfulAsset): Asset {
  return {
    id: entry.sys.id,
    title: entry.fields.title,
    url: fixHttps(fixImageUrl(entry.fields.file.url)),
  };
}

export function normalizeWebsite(
  ctx: RendrCtx,
  site: Entry<ContentfulWebsite>,
  normalizers: EntryNormalizer
): Website {
  return {
    id: site.sys.id,
    name: site.fields.name,
    path: site.fields.path,
    domains: site.fields.domains,
    culture: site.fields.culture,
    countryCode: site.fields.country_code,
    order: site.fields.order ? site.fields.order : 99,
    enabled: !!site.fields.enabled,
    cache: {
      sharedTtl: site.fields.shared_ttl ?? 0,
      ttl: site.fields.ttl ?? 0,
    },
    mainMenu: site.fields.main_menu ? site.fields.main_menu : {},
    settings: site.fields.settings ? site.fields.settings : {},
  };
}

export function createBlockDefinition(
  entry: Entry<BaseContentfulBlock>,
  type: string,
  settings: Settings
): BlockDefinition {
  return {
    id: entry.sys.id,
    container: "body", // default value for now
    order: 99, // default value for now
    settings: settings,
    type: type,
    meta: {},
  };
}
