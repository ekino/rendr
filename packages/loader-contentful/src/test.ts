import { Entry, ContentfulClientApi } from "contentful";
import { readFileSync } from "fs";

export function createDummyEntry<T>(data: T): Entry<T> {
  return {
    sys: {
      type: "Entry",
      id: "",
      createdAt: "",
      updatedAt: "",
      locale: "fr_FR",
      contentType: {
        sys: {
          type: "Link",
          linkType: "ContentType",
          id: "DummyEntry",
        },
      },
      space: {
        sys: {
          type: "Link",
          linkType: "Space",
          id: "DummySpace",
        },
      },
    },
    fields: data,
    toPlainObject: () => ({}),
    update: () => Promise.resolve(this),
  };
}

export function loadJson(path: string) {
  return JSON.parse(readFileSync(path, { encoding: "utf8" }));
}

export function getMockClient(
  methods: Partial<ContentfulClientApi>
): ContentfulClientApi {
  return {
    getAsset: async (id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getAssets: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getContentType: async (id: string) => {
      throw new Error("The method should not be called");
    },
    getContentTypes: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntries: async <T>(query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntry: async <T>(id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getSpace: async () => {
      throw new Error("The method should not be called");
    },
    getLocales: async () => {
      throw new Error("The method should not be called");
    },
    parseEntries: async <T>(raw: any) => {
      throw new Error("The method should not be called");
    },
    sync: async (query: any) => {
      throw new Error("The method should not be called");
    },
    ...methods,
  };
}
