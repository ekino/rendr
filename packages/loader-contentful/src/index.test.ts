import { createClient } from "contentful";
import { RequestCtx } from "@ekino/rendr-core";

import { createContentfulLoader } from "./index";
import { createNormalizer } from "./normalizer";
import { getMockClient, loadJson } from "./test";

function createContext(data: {}): RequestCtx {
  const ctx: RequestCtx = {
    // @ts-ignore
    req: jest.fn(),
    // @ts-ignore
    res: jest.fn(),
    hostname: "ekino.co.uk",
    pathname: "/",
    query: {},
    params: {},
    asPath: "/"
  };

  return {
    ...ctx,
    ...data
  };
}

describe("test loader", () => {
  it("page with no extend page", async () => {
    let call = 0;

    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        call++;

        if (call === 1) {
          expect(query).toEqual({
            content_type: "rendr_website",
            limit: 100,
            skip: 0
          });

          return loadJson(`${__dirname}/__fixtures__/loader/website.json`);
        }

        if (call === 2) {
          expect(query).toEqual({
            content_type: "rendr_page",
            "fields.path": "/",
            "fields.website.sys.id[in]": "632kl7enPots4PISSnD6DV",
            include: 10,
            limit: 1
          });

          return loadJson(`${__dirname}/__fixtures__/loader/main_page.json`);
        }

        if (call === 3) {
          expect(query).toEqual({
            content_type: "rendr_page",
            "fields.code": "root",
            "fields.website.sys.id[in]": "632kl7enPots4PISSnD6DV",
            include: 10,
            limit: 1
          });

          return loadJson(`${__dirname}/__fixtures__/loader/parent_page.json`);
        }
      }
    });

    const normalizer = createNormalizer();

    const clientFinder = (ctx: RequestCtx) => {
      return client;
    };

    const loader = createContentfulLoader(clientFinder, normalizer);

    const ctx = createContext({});

    const page = await loader(ctx);

    expect(page).toMatchSnapshot();
  });
});
