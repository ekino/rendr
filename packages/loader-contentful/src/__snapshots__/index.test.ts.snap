// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test loader page with no extend page 1`] = `
Page {
  "blocks": Array [
    Object {
      "container": "header",
      "order": 1,
      "settings": Object {
        "contents": "The header contents",
        "image": Object {
          "enabled": false,
          "id": "-",
          "title": "No picture",
          "url": "",
        },
        "image_position": "left",
        "mode": "standard",
        "subtitle": "",
        "title": "The header",
      },
      "type": "rendr.text",
    },
    Object {
      "container": "footer",
      "order": 1,
      "settings": Object {
        "contents": "The base footer",
        "image": Object {
          "enabled": false,
          "id": "-",
          "title": "No picture",
          "url": "",
        },
        "image_position": "left",
        "mode": "standard",
        "subtitle": "",
        "title": "The Footer",
      },
      "type": "rendr.text",
    },
    Object {
      "container": "article",
      "order": 1,
      "settings": Object {
        "contents": "All in!",
        "image": Object {
          "enabled": false,
          "id": "-",
          "title": "No picture",
          "url": "",
        },
        "image_position": "left",
        "mode": "standard",
        "subtitle": "",
        "title": "The page",
      },
      "type": "rendr.text",
    },
  ],
  "cache": Object {
    "ttl": 0,
  },
  "head": Object {
    "defaultTitle": "-",
    "htmlAttributes": Object {},
    "link": "",
    "meta": Array [
      Object {
        "content": "/",
        "name": "keywords",
      },
      Object {
        "content": "/",
        "name": "description",
      },
    ],
    "title": "homepage",
    "titleTemplate": "%s",
  },
  "id": "28dAkVDQuMmGBAX7IEOJDd",
  "path": "/",
  "settings": Object {},
  "statusCode": 200,
  "template": "default",
  "type": "document",
}
`;
